src/ddspdrum/parameter.py

```diff
@@ -0,0 +1,85 @@
+"""
+Parameter Class
+"""
+
+import numpy as np
+
+
+class Parameter:
+    """
+    Parameter class is a structure for keeping track of parameters
+    that have a specific range. Also handles functionality for converting
+    to and from a range between 0 and 1.
+
+    Parameters
+    ----------
+    value (float)   :   initial value of this parameter
+    minimum (float) :   minimum value that this parameter can take on
+    maximum (float) :   maximum value that this parameter can take on
+    scale   (float) :   scaling to apply when converting to and from a value with
+                        range [0,1]. Defaults to 1 which is linear scaling. A value
+                        less than 1 is a logarithmic relationship that fills more of the
+                        lower range of the parameter, whereas a scale value greater 1
+                        is an exponential relationship that fills more of the higher
+                        range of the parameter.
+    name    (str)   :   Optional name to give to this parameter.
+    """
+
+    def __init__(
+            self,
+            value: float,
+            minimum: float,
+            maximum: float,
+            scale: float = 1,
+            name: str = ""
+    ):
+        self.minimum = minimum
+        self.maximum = maximum
+        self.scale = scale
+        self.value = np.clip(value, minimum, maximum)
+        self.name = name
+
+    def __str__(self):
+        name = "{}, ".format(self.name) if self.name else ""
+        return "Parameter: {}Value: {}, Min: {}, Max: {}, Scale: {}".format(
+            name,
+            self.value,
+            self.minimum,
+            self.maximum,
+            self.scale
+        )
+
+    def set_value(self, new_value):
+        """
+        Set value of this parameter clipped to the minimum and maximum range
+
+        Parameters
+        ---------
+        new_value (float)   :   value to update parameter with
+        """
+        self.value = np.clip(new_value, self.minimum, self.maximum)
+
+    def set_value_0to1(self, new_value):
+        """
+        Set value of this parameter using a normalized value in the range [0,1]
+
+        Parameters
+        ----------
+        new_value (float)   :   value to update parameter with, in the range [0,1]
+        """
+        new_value = np.clip(new_value, 0, 1)
```

Comment:
@turian are we going to have gradient issues with clipping here?

Reply:
Yes, in general my preference would be that this module is not responsible for putting things in the correct range, and instead does an assert

Reply:
Cool, that's an easy fix.

Reply:
In terms of how you do ML when many of the parameters *should* be at the extrema of 0 or 1:
1) Pass the "preparameter" through a sigmoid-squashing function to turn it into a synth parameter. This is good IF you are okay with the the parameter getting close to 0 / 1 but never hitting it.
2) Pass the "preparameter" through a RELU-like squashing function to clip it, basically. The downside is you get no gradient (like dead neurons) so I think this is a no-go.
3) Add a steep penalty to the loss function if the parameter goes outside [0 - eps, 1 + eps], eps is maybe like 1e-4. Before you do a forward, the ML clips the parameters to 0, 1, *not the synth stuff*. Then it can get a gradient that is like: "Okay we want this parameter less than 0" from the synth but also know "okay but if I make this parameter further from 0 I get penalized so keep the parameter within eps of 0 and that sweet spot means the parameter is actually 0". This is the most promising approach I've tried.